File: frontend/apps/ios/Pantopus/Features/SavedPlaces/SavePlaceSheet.swift
//
//  SavePlaceSheet.swift
//  Pantopus
//
//  BLOCK 2E — the save/unsave affordance. Three pieces:
//    • `SaveBookmarkButton` — the bookmark toggle on Explore/Discover rows +
//      place detail (Frame 6): outline when not saved, filled when saved.
//    • `SavePlaceSheet` — the Save-place bottom sheet (Frame 7): a prefilled
//      Name field + a Home / Work / Other chooser + Save.
//    • `SavedPlacesStore` — a small @Observable cache the host (Explore) holds
//      so a place's saved-state reflects immediately and toggles wire to
//      `/api/saved-places` (POST upsert / DELETE with Undo).
//

import SwiftUI

// MARK: - Bookmark toggle (Frame 6)

/// The bookmark save toggle. Outline glyph when not saved, filled primary
/// disc when saved. Reflects state optimistically; the tap is handed up.
public struct SaveBookmarkButton: View {
    private let isSaved: Bool
    private let size: CGFloat
    private let onToggle: @MainActor () -> Void

    public init(isSaved: Bool, size: CGFloat = 34, onToggle: @escaping @MainActor () -> Void) {
        self.isSaved = isSaved
        self.size = size
        self.onToggle = onToggle
    }

    public var body: some View {
        Button(action: onToggle) {
            ZStack {
                Circle().fill(isSaved ? Theme.Color.primary600 : Theme.Color.appSurfaceSunken)
                Icon(
                    .bookmark,
                    size: size * 0.48,
                    strokeWidth: isSaved ? 2.4 : 2.0,
                    color: isSaved ? Theme.Color.appTextInverse : Theme.Color.appTextSecondary
                )
            }
            .frame(width: size, height: size)
            .contentShape(Circle())
        }
        .buttonStyle(.plain)
        .accessibilityLabel(isSaved ? "Saved" : "Save")
        .accessibilityIdentifier("savePlace.bookmarkToggle")
    }
}

// MARK: - Pending save payload

/// Seed for the Save-place sheet — the place the user is about to save,
/// prefilled from the row / detail it was triggered from.
public struct PendingSavePlace: Identifiable, Sendable, Hashable {
    public let id = UUID()
    public let label: String
    public let latitude: Double
    public let longitude: Double
    public let city: String?
    public let state: String?
    public let geocodePlaceId: String?
    public let sourceId: String?

    public init(
        label: String,
        latitude: Double,
        longitude: Double,
        city: String? = nil,
        state: String? = nil,
        geocodePlaceId: String? = nil,
        sourceId: String? = nil
    ) {
        self.label = label
        self.latitude = latitude
        self.longitude = longitude
        self.city = city
        self.state = state
        self.geocodePlaceId = geocodePlaceId
        self.sourceId = sourceId
    }
}

// MARK: - Save-place sheet (Frame 7)

public struct SavePlaceSheet: View {
    private let pending: PendingSavePlace
    private let onSave: @MainActor (_ label: String, _ choice: SavePlaceTypeChoice) -> Void
    private let onClose: @MainActor () -> Void

    @State private var name: String
    @State private var choice: SavePlaceTypeChoice = .other
    @FocusState private var nameFocused: Bool

    public init(
        pending: PendingSavePlace,
        onSave: @escaping @MainActor (String, SavePlaceTypeChoice) -> Void,
        onClose: @escaping @MainActor () -> Void
    ) {
        self.pending = pending
        self.onSave = onSave
        self.onClose = onClose
        _name = State(initialValue: pending.label)
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s4) {
            header
            nameField
            typePicker
            saveButton
            Spacer(minLength: 0)
        }
        .padding(.horizontal, Spacing.s5)
        .padding(.top, Spacing.s4)
        .padding(.bottom, Spacing.s5)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Theme.Color.appSurface)
        .presentationDetents([.height(420)])
        .presentationDragIndicator(.visible)
        .accessibilityIdentifier("savePlace.sheet")
    }

    private var header: some View {
        HStack {
            Text("Save place")
                .font(.system(size: 18, weight: .bold))
                .foregroundStyle(Theme.Color.appText)
            Spacer()
            Button(action: onClose) {
                Icon(.x, size: 18, color: Theme.Color.appTextSecondary)
                    .frame(width: 32, height: 32)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Close")
            .accessibilityIdentifier("savePlace.close")
        }
    }

    private var nameField: some View {
        VStack(alignment: .leading, spacing: Spacing.s2) {
            Text("Name")
                .font(.system(size: 12, weight: .semibold))
                .foregroundStyle(Theme.Color.appTextSecondary)
            HStack(spacing: Spacing.s2) {
                TextField("Name this place", text: $name)
                    .font(.system(size: 15, weight: .medium))
                    .foregroundStyle(Theme.Color.appText)
                    .focused($nameFocused)
                    .submitLabel(.done)
                    .accessibilityIdentifier("savePlace.nameField")
                Icon(.pencil, size: 15, color: Theme.Color.appTextMuted)
            }
            .padding(.horizontal, Spacing.s3)
            .frame(height: 46)
            .background(RoundedRectangle(cornerRadius: Radii.lg, style: .continuous).fill(Theme.Color.appSurfaceSunken))
            .overlay(
                RoundedRectangle(cornerRadius: Radii.lg, style: .continuous)
                    .stroke(nameFocused ? Theme.Color.primary300 : Theme.Color.appBorder, lineWidth: 1)
            )
        }
    }

    private var typePicker: some View {
        VStack(alignment: .leading, spacing: Spacing.s2) {
            Text("Type")
                .font(.system(size: 12, weight: .semibold))
                .foregroundStyle(Theme.Color.appTextSecondary)
            HStack(spacing: Spacing.s3) {
                ForEach(SavePlaceTypeChoice.allCases) { option in
                    typeOption(option)
                }
            }
        }
    }

    private func typeOption(_ option: SavePlaceTypeChoice) -> some View {
        let active = option == choice
        return Button { choice = option } label: {
            VStack(spacing: Spacing.s2) {
                ZStack {
                    RoundedRectangle(cornerRadius: Radii.lg, style: .continuous)
                        .fill(option.tileBackground)
                    Icon(option.icon, size: 18, color: option.tileForeground)
                }
                .frame(width: 44, height: 44)
                Text(option.label)
                    .font(.system(size: 12.5, weight: .semibold))
                    .foregroundStyle(active ? Theme.Color.appText : Theme.Color.appTextSecondary)
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, Spacing.s3)
            .background(
                RoundedRectangle(cornerRadius: Radii.lg, style: .continuous)
                    .fill(active ? Theme.Color.primary50 : Theme.Color.appSurface)
            )
            .overlay(
                RoundedRectangle(cornerRadius: Radii.lg, style: .continuous)
                    .stroke(active ? Theme.Color.primary600 : Theme.Color.appBorder, lineWidth: active ? 2 : 1)
            )
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier(option.accessibilityID)
    }

    private var saveButton: some View {
        Button {
            let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
            onSave(trimmed.isEmpty ? pending.label : trimmed, choice)
        } label: {
            HStack(spacing: Spacing.s2) {
                Icon(.bookmark, size: 17, strokeWidth: 2.4, color: Theme.Color.appTextInverse)
                Text("Save")
                    .font(.system(size: 15, weight: .bold))
                    .foregroundStyle(Theme.Color.appTextInverse)
            }
            .frame(maxWidth: .infinity)
            .frame(height: 50)
            .background(RoundedRectangle(cornerRadius: Radii.lg, style: .continuous).fill(Theme.Color.primary600))
        }
        .buttonStyle(.plain)
        .padding(.top, Spacing.s1)
        .accessibilityIdentifier("savePlace.saveBtn")
    }
}

// MARK: - Saved-places store (Explore integration)

/// A small cache the Explore host holds so a place's saved-state reflects
/// immediately. Mutations are optimistic and wire to `/api/saved-places`.
@Observable
@MainActor
public final class SavedPlacesStore {
    public private(set) var saved: [SavedPlaceDTO] = []

    /// Bound to the host's `.sheet(item:)` — the place pending a Save.
    public var pendingSave: PendingSavePlace?
    /// Bound to the host's Undo snackbar after an un-save.
    public var undo: SavedPlaceUndo?
    /// Transient error banner.
    public var toast: ToastMessage?

    private let api: APIClient
    private var loaded = false

    public init() {
        api = .shared
    }

    init(api: APIClient) {
        self.api = api
    }

    // MARK: Loading

    public func loadIfNeeded() async {
        guard !loaded else { return }
        await reload()
    }

    public func reload() async {
        do {
            let response: SavedPlacesListResponse = try await api.request(SavedPlacesEndpoints.list())
            saved = response.savedPlaces
            loaded = true
        } catch {
            // A failed background load just leaves toggles in their unsaved
            // default — the dedicated Saved-places screen surfaces the error.
        }
    }

    // MARK: Matching ("saved if it matches by geocode_place_id, else lat/lng")

    private static func matchKey(geocodePlaceId: String?, latitude: Double, longitude: Double) -> String {
        if let gid = geocodePlaceId, !gid.isEmpty { return "gid:\(gid)" }
        return String(format: "ll:%.5f,%.5f", latitude, longitude)
    }

    public func savedId(geocodePlaceId: String?, latitude: Double, longitude: Double) -> String? {
        let key = Self.matchKey(geocodePlaceId: geocodePlaceId, latitude: latitude, longitude: longitude)
        return saved.first { dto in
            Self.matchKey(geocodePlaceId: dto.geocodePlaceId, latitude: dto.latitude, longitude: dto.longitude) == key
        }?.id
    }

    public func isSaved(geocodePlaceId: String?, latitude: Double, longitude: Double) -> Bool {
        savedId(geocodePlaceId: geocodePlaceId, latitude: latitude, longitude: longitude) != nil
    }

    // MARK: Toggle

    /// Tapping the bookmark: saved → un-save (DELETE + Undo); not saved →
    /// open the Save sheet prefilled.
    public func toggle(_ pending: PendingSavePlace) {
        if let id = savedId(
            geocodePlaceId: pending.geocodePlaceId,
            latitude: pending.latitude,
            longitude: pending.longitude
        ) {
            Task { await remove(id: id) }
        } else {
            pendingSave = pending
        }
    }

    /// Commit the Save sheet — POST then optimistically insert the echo.
    public func commitSave(label: String, choice: SavePlaceTypeChoice) async {
        guard let pending = pendingSave else { return }
        pendingSave = nil
        let body = SavePlaceBody(
            label: label,
            placeType: choice.wire,
            latitude: pending.latitude,
            longitude: pending.longitude,
            city: pending.city,
            state: pending.state,
            geocodePlaceId: pending.geocodePlaceId,
            sourceId: pending.sourceId
        )
        // Optimistic insert so the toggle fills immediately.
        let optimistic = SavedPlaceDTO(
            id: "optimistic-\(pending.id.uuidString)",
            label: label,
            placeType: choice.wire,
            latitude: pending.latitude,
            longitude: pending.longitude,
            city: pending.city,
            state: pending.state,
            sourceId: pending.sourceId,
            geocodePlaceId: pending.geocodePlaceId,
            createdAt: nil
        )
        saved.append(optimistic)
        do {
            let response = try await api.request(
                SavedPlacesEndpoints.save(body),
                as: SavedPlaceResponse.self
            )
            if let i = saved.firstIndex(where: { $0.id == optimistic.id }) {
                saved[i] = response.savedPlace
            }
        } catch {
            saved.removeAll { $0.id == optimistic.id }
            toast = ToastMessage(text: "Couldn\u{2019}t save \(label).", kind: .error)
        }
    }

    /// Un-save — optimistic removal + DELETE, with an Undo that re-POSTs.
    public func remove(id: String) async {
        guard let index = saved.firstIndex(where: { $0.id == id }) else { return }
        let removed = saved[index]
        saved.remove(at: index)
        undo = SavedPlaceUndo(dto: removed, index: index)
        do {
            try await api.request(SavedPlacesEndpoints.remove(id: id))
        } catch {
            saved.insert(removed, at: min(index, saved.count))
            undo = nil
            toast = ToastMessage(text: "Couldn\u{2019}t remove \(removed.label).", kind: .error)
        }
    }

    public func undoRemove() async {
        guard let snapshot = undo else { return }
        undo = nil
        let dto = snapshot.dto
        saved.insert(dto, at: min(snapshot.index, saved.count))
        do {
            let response = try await api.request(
                SavedPlacesEndpoints.save(SavePlaceBody(from: dto)),
                as: SavedPlaceResponse.self
            )
            if let i = saved.firstIndex(where: { $0.id == dto.id }) {
                saved[i] = response.savedPlace
            }
        } catch {
            saved.removeAll { $0.id == dto.id }
            toast = ToastMessage(text: "Couldn\u{2019}t restore \(dto.label).", kind: .error)
        }
    }

    public func dismissUndo() {
        undo = nil
    }
}

#if DEBUG
#Preview("Save sheet") {
    Color.black.opacity(0.2)
        .sheet(isPresented: .constant(true)) {
            SavePlaceSheet(
                pending: PendingSavePlace(
                    label: "Blue Bottle Coffee",
                    latitude: 45.5152,
                    longitude: -122.6784,
                    city: "Portland",
                    state: "OR"
                ),
                onSave: { _, _ in },
                onClose: {}
            )
        }
}
#endif

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/saved_places/SavePlaceSheet.kt
@file:Suppress("MagicNumber", "PackageNaming")
@file:OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)

package app.pantopus.android.ui.screens.saved_places

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/**
 * BLOCK 2E Frame 6 — the bookmark save toggle on Explore/Discover rows + place
 * detail. Outline glyph when not saved, filled primary disc when saved.
 */
@Composable
fun SaveBookmarkButton(
    isSaved: Boolean,
    onToggle: () -> Unit,
    modifier: Modifier = Modifier,
    size: Dp = 34.dp,
) {
    Box(
        modifier =
            modifier
                .size(size)
                .clip(CircleShape)
                .background(if (isSaved) PantopusColors.primary600 else PantopusColors.appSurfaceSunken)
                .clickable(onClick = onToggle)
                .semantics {
                    contentDescription = if (isSaved) "Saved" else "Save"
                    role = Role.Button
                }
                .testTag("savePlace.bookmarkToggle"),
        contentAlignment = Alignment.Center,
    ) {
        PantopusIconImage(
            icon = PantopusIcon.Bookmark,
            contentDescription = null,
            size = size * 0.48f,
            strokeWidth = if (isSaved) 2.4f else 2.0f,
            tint = if (isSaved) PantopusColors.appTextInverse else PantopusColors.appTextSecondary,
        )
    }
}

/**
 * BLOCK 2E Frame 7 — the Save-place bottom sheet: a prefilled Name field + a
 * Home / Work / Other chooser + Save.
 */
@Composable
fun SavePlaceSheet(
    pending: PendingSavePlace,
    onSave: (label: String, choice: SavePlaceTypeChoice) -> Unit,
    onClose: () -> Unit,
) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    var name by rememberSaveable(pending) { mutableStateOf(pending.label) }
    var choice by remember(pending) { mutableStateOf(SavePlaceTypeChoice.Other) }

    ModalBottomSheet(onDismissRequest = onClose, sheetState = sheetState) {
        Column(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = Spacing.s5, vertical = Spacing.s2)
                    .padding(bottom = Spacing.s5)
                    .testTag("savePlace.sheet"),
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = "Save place",
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                    color = PantopusColors.appText,
                    modifier = Modifier.weight(1f),
                )
                Box(
                    modifier =
                        Modifier
                            .size(32.dp)
                            .clip(CircleShape)
                            .clickable(onClick = onClose)
                            .testTag("savePlace.close"),
                    contentAlignment = Alignment.Center,
                ) {
                    PantopusIconImage(
                        icon = PantopusIcon.X,
                        contentDescription = "Close",
                        size = 18.dp,
                        tint = PantopusColors.appTextSecondary,
                    )
                }
            }

            Spacer(Modifier.height(Spacing.s4))
            FieldLabel("Name")
            Spacer(Modifier.height(Spacing.s2))
            NameField(value = name, onValueChange = { name = it })

            Spacer(Modifier.height(Spacing.s4))
            FieldLabel("Type")
            Spacer(Modifier.height(Spacing.s2))
            Row(horizontalArrangement = Arrangement.spacedBy(Spacing.s3), modifier = Modifier.fillMaxWidth()) {
                SavePlaceTypeChoice.entries.forEach { option ->
                    TypeOption(option = option, active = option == choice, onClick = { choice = option }, modifier = Modifier.weight(1f))
                }
            }

            Spacer(Modifier.height(Spacing.s5))
            Row(
                modifier =
                    Modifier
                        .fillMaxWidth()
                        .height(50.dp)
                        .clip(RoundedCornerShape(Radii.lg))
                        .background(PantopusColors.primary600)
                        .clickable {
                            val trimmed = name.trim()
                            onSave(if (trimmed.isEmpty()) pending.label else trimmed, choice)
                        }
                        .testTag("savePlace.saveBtn"),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.Center,
            ) {
                PantopusIconImage(
                    icon = PantopusIcon.Bookmark,
                    contentDescription = null,
                    size = 17.dp,
                    strokeWidth = 2.4f,
                    tint = PantopusColors.appTextInverse,
                )
                Spacer(Modifier.size(Spacing.s2))
                Text("Save", fontSize = 15.sp, fontWeight = FontWeight.Bold, color = PantopusColors.appTextInverse)
            }
        }
    }
}

@Composable
private fun FieldLabel(text: String) {
    Text(text = text, fontSize = 12.sp, fontWeight = FontWeight.SemiBold, color = PantopusColors.appTextSecondary)
}

@Composable
private fun NameField(
    value: String,
    onValueChange: (String) -> Unit,
) {
    Row(
        modifier =
            Modifier
                .fillMaxWidth()
                .height(46.dp)
                .clip(RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurfaceSunken)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg))
                .padding(horizontal = Spacing.s3),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        Box(modifier = Modifier.weight(1f), contentAlignment = Alignment.CenterStart) {
            if (value.isEmpty()) {
                Text("Name this place", fontSize = 15.sp, color = PantopusColors.appTextMuted)
            }
            BasicTextField(
                value = value,
                onValueChange = onValueChange,
                singleLine = true,
                textStyle =
                    TextStyle(
                        fontSize = 15.sp,
                        fontWeight = FontWeight.Medium,
                        color = PantopusColors.appText,
                    ),
                cursorBrush = SolidColor(PantopusColors.primary600),
                modifier = Modifier.fillMaxWidth().testTag("savePlace.nameField"),
            )
        }
        PantopusIconImage(
            icon = PantopusIcon.Pencil,
            contentDescription = null,
            size = 15.dp,
            tint = PantopusColors.appTextMuted,
        )
    }
}

@Composable
private fun TypeOption(
    option: SavePlaceTypeChoice,
    active: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier =
            modifier
                .clip(RoundedCornerShape(Radii.lg))
                .background(if (active) PantopusColors.primary50 else PantopusColors.appSurface)
                .border(
                    width = if (active) 2.dp else 1.dp,
                    color = if (active) PantopusColors.primary600 else PantopusColors.appBorder,
                    shape = RoundedCornerShape(Radii.lg),
                )
                .clickable(onClick = onClick)
                .padding(vertical = Spacing.s3)
                .testTag(option.testTag),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        Box(
            modifier =
                Modifier
                    .size(44.dp)
                    .clip(RoundedCornerShape(Radii.lg))
                    .background(option.tileBackground),
            contentAlignment = Alignment.Center,
        ) {
            PantopusIconImage(icon = option.icon, contentDescription = null, size = 18.dp, tint = option.tileForeground)
        }
        Text(
            text = option.label,
            fontSize = 12.5.sp,
            fontWeight = FontWeight.SemiBold,
            color = if (active) PantopusColors.appText else PantopusColors.appTextSecondary,
        )
    }
}
